compiler/src/dotty/tools/dotc/core/NameOps.scala

```diff
@@ -80,15 +80,16 @@ object NameOps {
       case _ => false
 
     /** Is name a variable name? */
-    def isVariableName: Boolean = testSimple { n =>
-      n.length > 0 && {
-        val first = n.head
-        (((first.isLower && first.isLetter) || first == '_')
-          && (n != false_)
-          && (n != true_)
-          && (n != null_))
-      }
-    }
+    def isVariableName: Boolean =
+      testSimple { n =>
+        n.length > 0 && {
+          val first = n.head
+          (((first.isLower && first.isLetter) || first == '_')
+            && (n != false_)
+            && (n != true_)
+            && (n != null_))
+        }
+      } || name.is(PatMatQuoteTypeEv)
```

Comment:
Why this change is needed? This kind of specialization is not scalable.

Reply:
For a `$T` we need to create a name while typing for the `given Type[T] ` that will be available if the pattern matched. We need to use a semantic name to make sure that this name does not leak into userspace. This change makes this particular kind of semantic name be a valid pattern var to make typer handle it as such. 

I don't see any other case where we would use something like this and therefore I'm not sure we need it to scale. If we have another use-case in the future we can revise this and create a more general mechanism. We could achieve the same if we use an attachment on the `Ident` to mark such trees but this would come with some extra runtime overhead.

Reply:
Looking at it, `PatMatQuoteTypeEv` is not needed for anything specific quote wise. I can rename to `VarPattern` or something like that.

Reply:
Now it scales to any other synthetic var pattern.